### PR TITLE
Add a dummy OAuth2 authorization server

### PR DESCRIPTION
**Pull request by cocreature**
This is intended for use in the trigger service integration tests not
as an artifact that we ship to users.

For now the functionality is deliberately limited but as we need more in the tests, this might also be extended.

I did spend a bit of time searching for existing services for testing but couldn’t really find anything that I liked and this also served as a useful exercise for me to clarify some things and could serve as an easy reference for others.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      }
+      Query(params: _*)
+    }
+  }
+
+  // https://tools.ietf.org/html/rfc6749#section-4.1.3
+  case class Token(
+      grantType: String,
+      code: String,
+      redirectUri: Option[Uri],
```
Comment (aherrmann-da) [inline triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala:46]: https://tools.ietf.org/html/rfc6749#section-4.1.3 says it's required if present in the request, why is it optional here?

File: triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  // https://tools.ietf.org/html/rfc6749#section-4.1.2
+  case class Authorize(code: String, state: Option[String]) {
+    def toQuery: Query = state match {
+      case None => Query(("code", code))
+      case Some(state) => Query(("code", code), ("state", state))
+    }
+  }
+
+  // https://tools.ietf.org/html/rfc6749#section-4.1.1
```
Comment (aherrmann-da) [inline triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala:62]: Shouldn't this reference https://tools.ietf.org/html/rfc6749#section-5.1?

File: triggers/service/auth/src/test/scala/com/daml/oauth/server/Client.scala [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import com.daml.ports.Port
+import spray.json._
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+
+// This is a test client (using terminology from oauth) terminology.
```
Comment (aherrmann-da) [inline triggers/service/auth/src/test/scala/com/daml/oauth/server/Client.scala:20]: Double "terminology"

**Review by aherrmann-da — APPROVED** [review]
LGTM. A few questions.

File: triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala (hunk 1 above)
Comment (cocreature) [reply triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala:46]: Good point, that was when I still had it optional in the request as well which turned out to be annoying.

File: triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala (hunk 2 above)
Comment (cocreature) [reply triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala:62]: good catch!

**Pull request merged**